### PR TITLE
Modify permit for AA

### PR DESCRIPTION
**Pull request by luke7211**

### CODE REVIEW

**Comment by gbalabasquer** [comment]
It might be better to discuss it during the call tomorrow but I think I like more the idea to incorporate the code that we need directly in the Dai contract itself and nullify the use of libraries for it.
Also, we might want to only keep the original `permit` signature with the value broken in 3 variables than the other form. Not sure if the full signature in a bytes is becoming a new standard though.

File: packages/zksync-dai-bridge/contracts/l2/dai.sol [hunk 1]
```diff
@@ -18,6 +18,8 @@
 
 pragma solidity ^0.8.15;
 
+import {SignatureChecker} from "@matterlabs/signature-checker/contracts/SignatureChecker.sol";
```
Comment (kmbarry1) [inline packages/zksync-dai-bridge/contracts/l2/dai.sol:21]: I'd prefer to copy the code into this repo rather than having an external dependency.

Comment (kmbarry1) [reply packages/zksync-dai-bridge/contracts/l2/dai.sol:21]: Ah just saw Gonza's comment lol

File: packages/zksync-dai-bridge/contracts/l2/dai.sol [hunk 2]
```diff
@@ -202,16 +204,16 @@ contract Dai {
...
     uint256 value,
     uint256 deadline,
-    uint8 v,
-    bytes32 r,
-    bytes32 s
-  ) external {
+    bytes memory signature
+  ) public {
     require(block.timestamp <= deadline, "Dai/permit-expired");
+    require(owner != address(0), "Dai/invalid-permit");
```
Comment (kmbarry1) [inline packages/zksync-dai-bridge/contracts/l2/dai.sol:216]: Maybe this error message could be changed now?

File: packages/zksync-dai-bridge/contracts/l2/dai.sol (hunk 1 above)
Comment (talbaneth) [reply packages/zksync-dai-bridge/contracts/l2/dai.sol:21]: Just curious - why do we use matterlabs' code and not open-zeppelin's? are they the same?

Comment (luke7211) [reply packages/zksync-dai-bridge/contracts/l2/dai.sol:21]: Right now they are the same, but matterlabs indicated that they might change it in the future

**Comment by talbaneth** [comment]
> It might be better to discuss it during the call tomorrow but I think I like more the idea to incorporate the code that we need directly in the Dai contract itself and nullify the use of libraries for it. Also, we might want to only keep the original `permit` signature with the value broken in 3 variables than the other form. Not sure if the full signature in a bytes is becoming a new standard though.

regarding only keeping the signature broken to 3 - I assume we can't assume anything about the smart wallets signature scheme.

File: packages/zksync-dai-bridge/contracts/l2/dai.sol (hunk 2 above)
Comment (talbaneth) [reply packages/zksync-dai-bridge/contracts/l2/dai.sol:248]: @olivdb @luke7211 do we intend to change this?

**Comment by talbaneth** [comment]
I reviewed the recent changes in `dai.sol` since the discussion a few weeks ago (inline libs, rename internal function).
Updated code seems fine.

Note that OZ do have another check before `ecrecover` that we are missing - 
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/c22db8104e0eb4609c0f84a0546b2919dc6c0cb3/contracts/utils/cryptography/ECDSA.sol#L147
I assume it is fine not to have it as it is also missing from the mainnet dai contract, but let me know if you think otherwise.

**Comment by olivdb** [comment]
> Note that OZ do have another check before `ecrecover` that we are missing -
> https://github.com/OpenZeppelin/openzeppelin-contracts/blob/c22db8104e0eb4609c0f84a0546b2919dc6c0cb3/contracts/utils/cryptography/ECDSA.sol#L147
> I assume it is fine not to have it as it is also missing from the mainnet dai contract, but let me know if you think otherwise.

Yes, this check was intentionally left out. That's because signature malleability is acceptable here considering that each permit by a given owner burns a unique nonce.

**Pull request merged**